src/user-management/session.ts

```diff
@@ -0,0 +1,202 @@
+    this.cookiePassword = cookiePassword;
+    this.sessionData = sessionData;
+
+    const { clientId } = this.userManagement;
+
+    this.jwks = clientId
+      ? createRemoteJWKSet(new URL(userManagement.getJwksUrl(clientId)))
+      : undefined;
+  }
+
+  async authenticate(): Promise<
+    | AuthenticateWithSessionCookieSuccessResponse
+    | AuthenticateWithSessionCookieFailedResponse
+  > {
+    if (!this.sessionData) {
+      return {
+        authenticated: false,
+        reason:
+          AuthenticateWithSessionCookieFailureReason.NO_SESSION_COOKIE_PROVIDED,
+      };
+    }
+
+    const session =
+      await this.ironSessionProvider.unsealData<SessionCookieData>(
+        this.sessionData,
+        {
+          password: this.cookiePassword,
+        },
+      );
+
+    if (!session.accessToken) {
+      return {
+        authenticated: false,
+        reason:
+          AuthenticateWithSessionCookieFailureReason.INVALID_SESSION_COOKIE,
+      };
+    }
+
+    if (!(await this.isValidJwt(session.accessToken))) {
+      return {
+        authenticated: false,
+        reason: AuthenticateWithSessionCookieFailureReason.INVALID_JWT,
+      };
+    }
+
+    const {
+      sid: sessionId,
+      org_id: organizationId,
+      role,
+      permissions,
+    } = decodeJwt<AccessToken>(session.accessToken);
+
+    return {
+      authenticated: true,
+      sessionId,
+      organizationId,
+      role,
+      permissions,
+      user: session.user,
+      impersonator: session.impersonator,
+    };
+  }
+
+  async refresh(
+    options: RefreshOptions,
+  ): Promise<RefreshAndSealSessionDataResponse> {
+    if (options.sealSession && !options.cookiePassword) {
+      throw new Error('Cookie password is required for sealed sessions');
+    }
+
+    const session =
+      await this.ironSessionProvider.unsealData<SessionCookieData>(
+        this.sessionData,
+        {
+          password: options.cookiePassword as string,
+        },
+      );
+
+    if (!session.refreshToken || !session.user) {
+      return {
+        authenticated: false,
+        reason: RefreshAndSealSessionDataFailureReason.INVALID_SESSION_COOKIE,
+      };
+    }
+
+    const { org_id: organizationIdFromAccessToken } = decodeJwt<AccessToken>(
+      session.accessToken,
+    );
+
+    try {
+      const authenticationResponse =
+        await this.userManagement.authenticateWithRefreshToken({
+          clientId: this.userManagement.clientId as string,
+          refreshToken: session.refreshToken,
+          organizationId:
+            options.organizationId ?? organizationIdFromAccessToken,
+          session: {
+            // We want to store the new sealed session in this class instance, so this always needs to be true
+            sealSession: true,
+            cookiePassword: options.cookiePassword ?? this.cookiePassword,
+          },
+        });
+
+      if (options.sealSession) this.cookiePassword = options.cookiePassword;
```

Comment:
I'm strongly pro-curly braces for maintainability.
https://www.imperialviolet.org/2014/02/22/applebug.html
```suggestion
      if (options.sealSession) {
        this.cookiePassword = options.cookiePassword;
      }
```